### DISCUSSION TITLE
Analog to Digital Conversion: How to disable digital input buffer?

### DISCUSSION THREAD

**Discussion by btribley** [SpenceKonde/DxCore#92] [Q&A] [answered]
Using DxCore with AVR128DA28 no-bootloader, Arduino IDE.

I am having a devil of a time with stability on my ADC. Running a simple case with no digital pins changing states all is well. However, when running my for-real application the voltage at analog inputs drops inexplicably when certain digital ports change state. All port configuration is done once up-front before any ADC action is taken. I have carefully searched my code and I am not using analog pins in any digital operations including read.

The hardware checks out: No loose connections. 

Two questions:
1. Please confirm the proper way to refer to ADC input channels in analogRead(). I was using pin notation, so PD2 is PIN_PD2. The datasheet and ioavr128da28.h seem to use is by ADC channel number which means that pin PD2 is simply 2.  Pin notation seems to work, the straight numbering returns strange results. 
2. The datasheet says to disable digital input buffers for any analog input or unused pin. I don't see how to write to registers in plain Arduino or using DxCore.

**Comment by SpenceKonde** [comment, marked as answer]
It's done via PORTx.PINnCTRL register. - but the point of doing that is to reduce unnecessary power consumption on pins doing analog stuff or which are not in use and are floating (floating pins drain a huge amount of power in sleep mode on the modern AVRs. I swear it was never this bad on the classic AVRs. - but if you're trying to save power you do what you gotta do.)

To write registers.... you just write to them, like they were a variable:

```
// If I know that nothing was configured on PIN_PD0 and wanted tojust disable it:
PORTD.PIN0CTRL= PORT_ISC_INPUT_DISABLE_gc;;
// that assumes that nothing else is set on that register, or if it is, that we want to unset it. A more cautious approach would be:
PORTD.PIN0CTRL = (PORTD.PIN0CTRL & ~PORT_ISCgm) | PORT_ISC_INPUT_DISABLE_gc
// And yes, you do basically need the datasheet open in one window, Arduino in another, and the IO header file in a third, in order to write code like that. 
// second line, we first bitwise AND the current value with the bitwise inverted Input Sense Configuration Group mask definition provided by the IO headers. Then we bitwise OR it with the group code for the setting we want (_gm and _gc stand for group mask, group code. Just like there are also _bm defines all over the place: bit mask.
```

Pins have ports (x) and bits (n) for Pxn, Analog channels have numbers. Microchip did not make any attempt at making thus make any sense beyond "pin D1 is connected to analog channel 1" I took one look at the mess, and looked back at what it turned into on classic AVR cores and made the same decision I did for megaTinyCore: analogRead() expects Arduino pin numbers, and I asked  the pinout chart maker (@freemovers on megaTinyCore and @MCUdude  here and on ATTinyCore)  to not list the analog channel numbers at all, and we can then end one of the most common 

In megaTinyCore and DxCore, (and after ATTinyCore 2.0.0 ios out, on there too)  you should be using the Arduino pin numbers or (preferably) rhe PIN_Pxn  notation),. The core is very capable of looking up analog channels as needed. 
I didn't think it had an adverse impact on analog performance to leave the digital input buffers on. I think there is something ELSE going on that you are unaware of which is causing these readings to fluctuate. That's my theory,. Maybe your power supply is weak, and some powerhog you have connected is flinging the supply rail around. Maybe it's insufficiently decoupled, maybe there's a loose connection and something that you think is powered is only barely alive because of leakage through the protection diodes... something like that 
:

**Comment by btribley** [comment]
Thanks for  your explanation of how to reference the pin registers. Regarding the problem I was seeing, you were right. Nothing to do with hardware, all supply voltages and connections good. I am using a PCB. I was not reading EEPROM correctly and depending on what buttons were pushed a different calibration value was read. All OK now. Thanks for your help and have a good day!
